### ISSUE TITLE
Very long shader/pipeline compilation

### ISSUE THREAD

**Issue by eldidou** [godotengine/godot#99964]
### Tested versions

Reproducible: 4.4-dev4, 4.4-dev5
Not Reproducible: <= 4.4-dev3

### System information

Godot v4.4.dev5 - Ubuntu 24.04.1 LTS 24.04 on X11 - X11 display driver, Multi-window, 2 monitors - Vulkan (Forward+) - integrated Intel(R) UHD Graphics (CML GT2) - Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz (12 threads)

### Issue description

Starting from 4.4-dev4, shader (or pipeline?) compilation takes a very long time, around 6s, sometimes 12s.
Before it was just a small stutter, ~250ms maybe.

To reproduced it reliably, I've tried to remove shader caches from different places.

- ~/.local/share/godot/app_userdata/my_godot_project/shader_cache/
- ~/.local/share/godot/app_userdata/my_godot_project/vulkan/
- ~/.local/share/godot/shader_cache/
- ~/my_godot_project/.godot/shader_cache/
- ~/.cache/godot/
- ~/.cache/mesa_shader_cache/

The one that triggers the long freeze when cleaned is `~/.cache/mesa_shader_cache/`.

It can happen in many different situations:

- at first display of a material
- scene loading
- project loading in the editor
- scene opening in the editor
- selecting a mesh instance in the editor
- changing a parameter of a material in the editor

When it happens on first display of the material (for example by setting a mesh material to `StandardMaterial3D.new()` by script), it's possible to monitor it: it shows 1 additional Compilation Specialization and 1 additional Compilation Surface. (circled in red in the screenshot below)
Subsequent displays of another similar material will also show 1 additional Compilation Specialization and 1 additional Compilation Surface, but will not have a long freeze.

![Image](https://github.com/user-attachments/assets/ba55354a-57ff-4468-864a-9cec634ff6f9)

![Image](https://github.com/user-attachments/assets/e2421675-d450-41dd-abd5-644b7c61173a)

![Image](https://github.com/user-attachments/assets/e3f88fb4-90b0-451c-92f0-48c399d48152)

### Steps to reproduce

1. Remove shader caches (`~/.cache/mesa_shader_cache/` in my case)
2. run a scene with a mesh
3. change the material to `StandardMaterial3D.new()`

### Minimal reproduction project (MRP)

[shader_compilation.zip](https://github.com/user-attachments/files/17994403/shader_compilation.zip)

**Comment by Chaosus** [comment]
Looks like a duplicate of https://github.com/godotengine/godot/issues/99420

**Comment by clayjohn** [comment]
> Both are probably related to the changes on rendering released in 4.4-dev4, but (from what I read) [#99420](https://github.com/godotengine/godot/issues/99420) seems to be only 2d, and for each frame, while here it seems only 3d, and only once, usually first display of a material/shader.

That's right. I don't think #99420 is related. This is most likely a regression from the Ubershader PR (https://github.com/godotengine/godot/pull/90400)

CC @DarioSamo

**Comment by eldidou** [comment]
I've bisected the issue to https://github.com/godotengine/godot/commit/427ba09efc70b2ad836bfa2441a242bbda10ba36, in particular in _servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl_, the mask size changes (from `15U` to `63U`) seems to trigger the issue.

**Comment by DarioSamo** [comment]
> I've bisected the issue to [427ba09](https://github.com/godotengine/godot/commit/427ba09efc70b2ad836bfa2441a242bbda10ba36), in particular in _servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl_, the mask size changes (from `15U` to `63U`) seems to trigger the issue.

I think this might be something along the lines of the shader compiler unrolling as many times as it could possibly have it instead of making a dynamic loop like it should, because increasing the mask size itself shouldn't be _that_ big of an issue.

**Comment by DarioSamo** [comment]
Could you do me a favor and test this branch?

https://github.com/DarioSamo/godot/tree/dont-unroll-ubershader

**Comment by eldidou** [comment]
Yes, https://github.com/DarioSamo/godot/tree/dont-unroll-ubershader works. 👍
With it, shader compilation is back to around few 100ms, compared to ~6s without.

**Comment by clayjohn** [comment]
On my device I get:
```
//before 427ba09
Updated PSO cache (3.3 MiB)
2113.838

//after 427ba09
Updated PSO cache (10.0 MiB)
6636.551

// with no-unroll
Updated PSO cache (6.1 MiB)
1158.878
```

**Comment by clayjohn** [comment]
@DarioSamo I tested out my idea for further reducing the compile time by moving the soft shadow settings to proper specialization constants (even when using the ubershader). It worked, but no better than using unroll, and when combined with unroll, it didn't make it any faster. So I think it is safe to just go with the unrolling PR. 

Here is the commit for posterity https://github.com/clayjohn/godot/commit/2e07ca83dc3f63cf447fc5ab8fca3c8fbada2d16
